File: restaurants/canyon/get_menu.py
from urllib import request
from bs4 import BeautifulSoup
import json
import datetime

from os import path
directory = path.dirname(path.realpath(__file__))

titles = [
    "KÖTT",
    "FISK",
    "VEGETARISKT",
    "ASIATISKT",
    "TACOS"
]

def main():
    html = request.urlopen("http://restaurang-einstein.se").read().decode()

    html_tree = BeautifulSoup(html, features="html.parser")

    relevant_html = html_tree.find_all(class_="content-wrapper")[3].contents

    now = datetime.datetime.now().isocalendar()

    menu = {
        "restaurant": "Café Canyon",
        "year": now[0],
        "week": now[1],
        "days": {}
    }

    for day in range(0, 5):
        i = day * 4 + 3

        dishes = []
        
        try:
            html_dishes = relevant_html[i].contents[1].contents[0].contents
            i = 0
            for html_dish in html_dishes:
                dish = html_dish.string
                if dish != None:
                    dish = dish.strip()
                    if dish != "":
                        title = ""
                        if i < len(titles):
                            title = titles[i]
                        dishes.append({
                            "title": title,
                            "description": dish
                        })
                i += 1
        except:
            dishes = ["N/A"]

        menu["days"][day] = dishes
    
    print(json.dumps(menu, ensure_ascii=False))
main()

File: restaurants/canyon/getMenu.js
const cheerio = require("cheerio")
const http = require("http")

function getHtml() {
    let options = {
        hostname: 'restaurang-einstein.se',
        port: 80,
        method: "GET"
    }
    
    return new Promise((resolve, reject) => {
        let req = http.request(options, (res) => {
            let data = ""
            res.on("data", d => data += d)
            res.on("end", () => {
                resolve(data)
            })
        })
        
        req.end()
    })
}


module.exports = async () => {
    let html = await getHtml()
    const $ = cheerio.load(html)

    let relevantHtml = $(".content-wrapper")[3].children

    let now = new Date()

    let menu = {
        restaurant: "Café Canyon",
        year: now.getFullYear(),
        week: getWeek(now),
        days: {}
    }

    for (let day = 0; day < 5; day++) {
        let i = day * 4 + 3

        let dishes = []

        try {
            let htmlDishes = relevantHtml[i].children[1].children[0].children
            htmlDishes.forEach(htmlDish => {
                let dish
                try {
                    dish = htmlDish.children[0].data
                } catch (e) {
                    return
                }

                if (!dish) return

                dish = dish.trim()
                if (dish === "") return

                dishes.push({
                    title: "",
                    description: dish
                })
            })
        } catch (e) {
            // Do nothing
        }
        menu.days[day] = dishes
    }

    return menu
}

function getWeek(date) {
    var year = date.getFullYear();
    var oneJan = new Date("Jan 1 " + year);
    var oneJanDay = (oneJan.getDay() + 6) % 6 
    var day = Math.floor((date.getTime() - oneJan.getTime()) / 86400000);
    var week = Math.floor(day / 7);

    if (oneJanDay < 4) {
        week++;
    }
    return week;
}